File: jupyter2/problems/oop_basics/subtyping/go/subtyping.go
package main
import (
	"fmt"
	"math"
)

/** begin my answer */

/** end my answer */
func veryClose(x, y float64) bool {
	return math.Abs(x - y) < 1e-6
}

func main() {
	e0 := mkFreeEllipse(16., 5.)
	r0 := mkFreeRect(16., 5.)
	e1 := mkFreeEllipse( 6., 3.)
	r1 := mkFreeRect( 6., 3.)
	d0 := mkEllipse(16., 5., -15., -4.)
	q0 := mkRect(16., 5., -15., -4.)
	d1 := mkEllipse( 6., 3.,  -3., -2.)
	q1 := mkRect( 6., 3.,  -3., -2.)
	if ! veryClose(e0.area(), 251.327412) { panic("wrong") }
	if ! veryClose(r0.area(), 80.)        { panic("wrong") }
	if ! veryClose(e1.area(), 56.548668)  { panic("wrong") }
	if ! veryClose(r1.area(), 18.)        { panic("wrong") }
	if ! veryClose(d0.area(), 251.327412) { panic("wrong") }
	if ! veryClose(q0.area(), 80.)        { panic("wrong") }
	if ! veryClose(d1.area(), 56.548668)  { panic("wrong") }
	if ! veryClose(q1.area(), 18.)        { panic("wrong") }
	if !(smaller(e0, e1) == 1) { panic("wrong") }
	if !(smaller(e0, r0) == 1) { panic("wrong") }
	if !(smaller(d1, q0) == 0) { panic("wrong") }
	if !(smaller(e0, d0) == 0) { panic("wrong") }
	s0 := mkFreeShapes()
	s1 := mkShapes()
	s2 := mkMixedShapes()
	// un-comment only lines below that run successfully
	// if ! (countContains(s0, 0.0, 0.0) == 3) { panic("wrong") }
	if ! (countContains(s1, 0.0, 0.0) == 3) { panic("wrong") }
	// if ! (countContains(s2, 0.0, 0.0) == 6) { panic("wrong") }
	if ! veryClose(sumArea(s0), 405.876080) { panic("wrong") }
	// if ! veryClose(sumArea(s1), 405.876080) { panic("wrong") }
	if ! veryClose(sumArea(s2), 811.752160) { panic("wrong") }
	fmt.Println("OK")
}

File: jupyter2/problems/oop_basics/subtyping/ml/subtyping.ml
(** begin my answer *)

(** end my answer *)

let very_close x y =
  abs_float (x -. y) < 1e-6
;;

let main () =
  let e0 = mk_free_ellipse 16. 5. in
  let r0 = mk_free_rect    16. 5. in
  let e1 = mk_free_ellipse  6. 3. in
  let r1 = mk_free_rect     6. 3. in
  let d0 = mk_ellipse 16. 5. (-15.) (-4.) in
  let q0 = mk_rect    16. 5. (-15.) (-4.) in
  let d1 = mk_ellipse  6. 3. (-3.)  (-2.) in
  let q1 = mk_rect     6. 3. (-3.)  (-2.) in
  assert (very_close e0#area 251.327412);
  assert (very_close r0#area 80.);
  assert (very_close e1#area 56.548668);
  assert (very_close r1#area 18.);
  assert (very_close d0#area 251.327412);
  assert (very_close q0#area 80.);
  assert (very_close d1#area 56.548668);
  assert (very_close q1#area 18.);
  assert (smaller e0 e1 = 1);
  assert (smaller e0 r0 = 1);
  assert (smaller d1 q0 = 0);
  assert (smaller e0 d0 = 0);
  let s0 = mk_free_shapes () in
  let s1 = mk_shapes () in
  let s2 = mk_mixed_shapes () in
  (* un-comment lines below that pass static type checking *)
  (* assert (count_contains s0 0. 0. = 3); *)
  assert (count_contains s1 0. 0. = 3);
  (* assert (count_contains s2 0. 0. = 6); *)
  assert (very_close (sum_area s0) 405.876080);
  assert (very_close (sum_area s1) 405.876080);
  assert (very_close (sum_area s2) 811.752160);
  Printf.printf "OK\n"
;;

main()
